src/docmanager/context.ts

```diff
@@ -0,0 +1,511 @@
+   *
+   * #### Notes
+   * This is a read-only property.  The model will have an
+   * empty `contents` field.
+   */
+  get contentsModel(): IContentsModel {
+    return this._manager.getContentsModel(this._id);
+  }
+
+  /**
+   * Get the kernel spec information.
+   *
+   * #### Notes
+   * This is a read-only property.
+   */
+  get kernelSpecs(): IKernelSpecIds {
+    return this._manager.getKernelSpecs();
+  }
+
+  /**
+   * Get whether the context has been disposed.
+   */
+  get isDisposed(): boolean {
+    return this._manager === null;
+  }
+
+  /**
+   * Dispose of the resources held by the context.
+   */
+  dispose(): void {
+    if (this.isDisposed) {
+      return;
+    }
+    this._manager = null;
+    this._id = '';
+  }
+
+  /**
+   * Change the current kernel associated with the document.
+   */
+  changeKernel(options: IKernelId): Promise<IKernel> {
+    return this._manager.changeKernel(this._id, options);
+  }
+
+  /**
+   * Save the document contents to disk.
+   */
+  save(): Promise<void> {
+    return this._manager.save(this._id);
+  }
+
+  /**
+   * Revert the document contents to disk contents.
+   */
+  revert(): Promise<void> {
+    return this._manager.revert(this._id);
+  }
+
+  /**
+   * Get the list of running sessions.
+   */
+  listSessions(): Promise<ISessionId[]> {
+    return this._manager.listSessions();
+  }
+
+  /**
+   * Add a sibling widget to the document manager.
+   *
+   * @param widget - The widget to add to the document manager.
+   *
+   * @returns A disposable used to remove the sibling if desired.
+   *
+   * #### Notes
+   * It is assumed that the widget has the same model and context
+   * as the original widget.
+   */
+  addSibling(widget: Widget): IDisposable {
+    return this._manager.addSibling(this._id, widget);
+  }
+
+  private _id = '';
+  private _manager: ContextManager = null;
+}
+
+
+
+/**
+ * A namespace for module private data.
+ */
+namespace Private {
+
+  /**
+   * A signal emitted when the kernel changes.
+   */
+  export
+  const kernelChangedSignal = new Signal<IDocumentContext, IKernel>();
+
+  /**
+   * A signal emitted when the path changes.
+   */
+  export
+  const pathChangedSignal = new Signal<IDocumentContext, string>();
+
+  /**
+   * A signal emitted when the model is saved or reverted.
+   */
+  export
+  const dirtyClearedSignal = new Signal<IDocumentContext, void>();
+
+}
+
+
+/**
+ * An object which manages the active contexts.
+ */
+export
+class ContextManager implements IDisposable {
+  /**
+   * Construct a new context manager.
+   */
+  constructor(contentsManager: IContentsManager, sessionManager: INotebookSessionManager, opener: (id: string, widget: Widget) => IDisposable) {
+    this._contentsManager = contentsManager;
+    this._sessionManager = sessionManager;
+    this._opener = opener;
+    // Fetch and store the kernelspecids.
+    sessionManager.getSpecs().then(specs => {
+      this._kernelspecids = specs;
```

Comment:
I like that. Fetching the kernel specs is part of the application initialization, so the plugin creating the context manager can return a promise that fulfills when the kernelspecs are fetched and then the context manager is created.